libs/tiled-bitmap/src/layeroperations.cc

```diff
@@ -209,6 +209,52 @@ void CommonOperations::draw(cairo_t* cr, const ConstTile::Ptr tile,
   }
 }
 
+PipetteLayerOperations::PipetteColor PipetteCommonOperations::sumPixelValues(Scroom::Utils::Rectangle<int> area, const ConstTile::Ptr tile, int bps, int spp)
```

Comment:
If you are going to provide different implementation for `spp==4` and `spp==3`, then perhaps they should be in different classes. After all, they share virtually no code.

On the other hand, the only real difference between the two cases is the names of the samples. If `sumPixelValues()` were to return a vector of samples, instead of a map, then they would be identical.

Only problem left is to find a suitable location to convert the vector into a map :wink:

Reply:
I think for myself, i would prefer to split them into 2 classes. `PipetteCommonOperationsCMYK` and `PipetteCommonOperationsRGB` would be options imo. The specific operations such as `OperationsCMYK24` can then inherit from one of the two above (in this case being `PipetteCommonOperationsCMYK`). The spp would then be implicit, the bps *could* be calculated from the `getBpp()` function. (I think you noticed that im not a huge fan of vectors for this :smile: ). My only worry for this approach would be the class hierarchy and large inheritance chain that we would end up with for `OperationsCMYKx` and `Operationsxbpp`

Reply:
After all edits, the plan seems sound :wink: 

For future reference: You were asking me to review concrete `OperationCMYKx` classes, where some of the methods new the bpp and spp values they were supposed to be operating with, and some other methods were not. For me, that's a sign of bad design, and that was the trigger to start this discussion.

Now, a long inheritance hierarchy is "merely" a code smell. So although the concern is valid, introducing bad design to avoid a smell is, eh, misguided.

Having said all that, the inheritance tree we're creating here is still pretty much ok (imho, and stuff). Yes, there are a fair number of classes, but they have clear responsibilities, and all method implementations match the responsibilities of their respective classes. Furthermore, you'll notice that each method on the interface is implemented in exactly one location in the inheritance tree. There are no overrides, there are no calls to parent methods in base classes. We could do much worse. In fact, this tree is fairly easy to clean up, if the need arises. Maybe by moving from inheritance to composition. Maybe by splitting the `LayerOperations` interface further. There are several possibilities to choose from.